sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,305 @@
+
+#include "sw/device/lib/dif/dif_padctrl.h"
+
+// Generated.
+// Note that if the the number of DIO or MIO pads is such that all attribute can
+// fit in one register the PADCTRL_DIO_PADS0_REG_OFFSET and
+// PADCTRL_MIO_PADS0_REG_OFFSET #defines from padctrl_regs.h will no longer
+// contain '0' and become PADCTRL_DIO_PADS_REG_OFFSET and
+// PADCTRL_MIO_PADS_REG_OFFSET respectively which will require altering their
+// names in the source below.
+#include "padctrl_regs.h"
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+
+_Static_assert(PADCTRL_PARAM_NDIOPADS == 15,
+               "Padctrl instantiation parameters have changed");
+_Static_assert(PADCTRL_PARAM_NMIOPADS == 32,
+               "Padctrl instantiation parameters have changed");
+_Static_assert(PADCTRL_PARAM_ATTRDW == 8,
+               "Padctrl instantiation parameters have changed");
+_Static_assert(PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG ==
+                   PADCTRL_DIO_PADS_ATTR_FIELDS_PER_REG,
+               "Padctrl instantiation parameters have changed");
+
+static const int kPadctrlRegisterWidthInBytes = (PADCTRL_PARAM_REG_WIDTH / 8);
+
+static const uint32_t kPadctrlAttrFieldMask = PADCTRL_MIO_PADS0_ATTR0_MASK;
+
+/**
+ * Checks if an attribute field contains conflicting attributes
+ */
+static bool conflicting_attributes(dif_padctrl_attr_t attrs) {
+  // Cannot be both pull up and pull down
+  if ((attrs & kDifPadctrlAttrPullDown) && (attrs & kDifPadctrlAttrPullUp)) {
+    return true;
+  }
+
+  // Cannot combine pull down with open drain
+  if ((attrs & kDifPadctrlAttrPullDown) && (attrs & kDifPadctrlAttrOpenDrain)) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Check if a given pad `kind` and `index` refers to a valid pad
+ */
+static bool is_pad_index_valid(dif_padctrl_pad_kind_t kind,
+                               dif_padctrl_pad_index_t index) {
+  if (kind == kDifPadctrlPadMio) {
+    return index < PADCTRL_PARAM_NMIOPADS;
+  } else {
+    return index < PADCTRL_PARAM_NDIOPADS;
+  }
+}
+
+/**
+ * Check any bit set in `attrs` corresponds to a valid attribute
+ */
+static bool attrs_are_valid(dif_padctrl_attr_t attrs) {
+  return (attrs & ~kPadctrlAttrFieldMask) == 0;
+}
+
+/**
+ * Calculate the `bitfield_field32_t` needed to access a pad's attribute field.
+ * Assumes the pad is a valid one.
+ */
+static bitfield_field32_t calc_pad_attr_field(dif_padctrl_pad_kind_t kind,
+                                              dif_padctrl_pad_index_t index) {
+  bitfield_field32_t field;
+
+  field.index =
+      (index % PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG) * PADCTRL_PARAM_ATTRDW;
+  field.mask = kPadctrlAttrFieldMask;
+
+  return field;
+}
+
+/**
+ * Calculate the register offset for a pad's attribute register. Assumes the pad
+ * is a valid one.
+ */
+static ptrdiff_t calc_pad_attr_reg_offset(dif_padctrl_pad_kind_t kind,
+                                          dif_padctrl_pad_index_t index) {
+  uint32_t register_index;
```

Comment:
I'd avoid pre-declaring variables like this. Move `register_index` to where its first used.

I'd also write code that conditionally initializes a variable as
```c
foo_t foo;
if (cond) {
  foo = blah;
} else {
  ...
}
```
to make the initialization relationship clearer (note the lack of a newline).